### PR TITLE
Document the public api

### PR DESCRIPTION
**Pull request by MSleepyPanda**
While working with the API, i couldn't help but notice the lack of documentation. I talked with @glennw on IRC and he explained me a lot of details which i wrote down, so that new contributors will learn from my mistakes. I also tried to explain the relationship between the different parts of the API.

It's still a bit rough but it's something.
There's an equivalent PR for webrender_traits as well.

### CODE REVIEW

File: src/renderer.rs [hunk 1]
```diff
@@ -1877,4 +1915,3 @@ struct VertexBufferAndOffset {
     buffer: VertexBuffer,
     offset: u32,
 }
-
```
Comment (KiChjang) [inline src/renderer.rs]: nit: Re-add this newline.

**Comment by emilio** [comment]
Thanks for doing this! It's awesome! I have skimmed over it just a bit, and it looks good to me, but Glenn will have to sanity-check and sign-off this.

r? @glennw 

Thanks again for doing this! :)

**Comment by MSleepyPanda** [comment]
Thanks for the encouraging words!

If you'd like coverage on other parts of the api, please let me know and i'll wrap my mind around it :) Personally, the pipeline part still bugs me a bit, but i'm getting closer

**Comment by glennw** [comment]
@MSleepyPanda Fantastic, thanks for taking the time to do this!

**Pull request merged**